[DRAFT] OGC Features API (wfs3) endpoint

olivierdalang opened the pull request:
This is the WIP implementation for WFS3 (aka OGC Features API) for Django applied to Geocity.

Please see `django_wfs3/README.md` for more info on implementation.

![geocitywfs3](https://user-images.githubusercontent.com/1894106/135130336-70a7945b-b676-44ce-a490-52cc41181f03.gif)

It's not ready yet, but opening this PR already so you can review and see if it takes the desired direction.

Roadmap :
- [x] basic implementation
- [x] working readonly endpoints 
  - [x] PermitRequestViewSet (incl. subclasses for point/line/poly only geometries so they can load in QGIS)
  - [x] PermitRequestGeoTimeViewSet
- [ ] fix CRS in QGIS (seems CRS is simply ignored - do we have a working example, or is it an issue with QGIS ?)
- [x] various other fixes (see TODO in code)
- [x] test with oauth2 (it should work already, but unsure how this must be configured from QGIS)

monodo commented:
If environment variable PREFIX_URL has a value, it is missing in the first link:

![image](https://user-images.githubusercontent.com/3356536/135845205-b5ea7595-7fbf-4a9c-b8d1-c16dc785b2a5.png)

olivierdalang commented:
> If environment variable PREFIX_URL has a value, it is missing in the first link:

Good catch, should be fixed.

olivierdalang commented:
Made the changes as discussed :
- removed geotime endpoints
- added permits endpoints with specific geometries extracts

![image](https://user-images.githubusercontent.com/1894106/136019050-88da2a3f-a4df-420a-8fe8-fe9e95781b3f.png)

monodo commented:
please go into web container and execute `black .`

olivierdalang commented:
Added a new commit that makes layer homogeneous, which hopefully will fix loading of prod data in QGIS. This changes a bit the behavior, so please review carefully the following changes :

- We now return null instead of an empty point when there's no geometry. Unsure why it used to be an empty point  ? Depending on how/where this is used, it could break existing behavior.

- We now add a small padding to bounding boxes, to ensure it returns a polygon (BOX2Ds representation are not necessarily polygons). This changes the representation, IMO it's even more clear with the small padding. Currently set to 1m, we could also make that very low if we don't want a visible padding.

![image](https://user-images.githubusercontent.com/1894106/136183734-3bc800b6-7708-4d28-9770-d4ef6649e2a6.png)

monodo commented on django_wfs3/README.md:
```diff
@@ -0,0 +1,58 @@
+
+This is probably still relatively far from a full OGC Services API implementation and currently only aims to support read-only view from QGIS.
+
+This app will as some point be factored out
```
typo "at"

monodo commented on geomapshark/settings.py:
```diff
@@ -370,5 +370,11 @@
         "rest_framework.authentication.SessionAuthentication",
-    )
+        # TODO : remove this once OAuth2 settings from qgis are clear
+        "rest_framework.authentication.BasicAuthentication",
```
oAuth2 with QGIS has been tested and workd fine, thus, you can remove this line @olivierdalang

monodo reviewed (commented):
I'll push this little directly in your branch and merge right ahead
